include/etcpal/cpp/queue.h

```diff
@@ -103,8 +103,14 @@ class Queue
   template <class Rep, class Period>
   bool Receive(T& data, const std::chrono::duration<Rep, Period>& timeout);
   bool ReceiveFromIsr(T& data);
+  bool Reset();
   bool IsEmpty() const;
   bool IsEmptyFromIsr() const;
+  bool IsFull() const;
+  bool IsFullFromIsr() const;
+  unsigned int Size() const;
+  unsigned int SizeFromIsr() const;
+  unsigned int Available() const;

```

Comment:
I think we should tweak these function names a bit so that people can grok them more quickly. The main issue I think is that "size" is used both at construction/creation to specify the _maximum_ size, and during use to get the _used_ size. So, I think there are two main ways we could do with the API:

1. Change `size` at construction (and creation in the C API) to be `max_size`. Add a private member here for the max size and add a constexpr getter `MaxSize()` for it. Then keep `Size()` and `SizeFromIsr()` but remove `Available()` (you can get that just by doing `MaxSize()` - `Size()`).

2. Keep `size` the same in the constructor, but change `Size()` to `NumUsed()` and `Available()` to `NumAvailable()` for better symmetry.

I will leave it up to you which you think is better. Obviously we would need to translate these to the C API as well.

Reply:
method 1 seems preferable to me, as std::queue uses size for number of elements

Reply:
We can definitely expose it if you think we should. There's no rule that says EtcPal needs to expose every API function for a given FreeRTOS module (and keep in mind we target more platforms than just FreeRTOS), so it should be a decision made based on the design and ergonomics of the platform-neutral EtcPal API. That being said, I would definitely hear an argument saying that getting the number of queue slots available is a good convenience function to have.

I am not a big fan of the name Available(), for two reasons:

1. It's not clear - what does this mean? `if (queue.Available() > 0)` <- to me that reads as checking whether the queue is available, whatever that means.
2. Since Available and Size are opposites in this API design, they should have opposite/complementary names. For example, NumUsed and NumAvailable.

If we want to keep Available as a concept, I would lean toward option 2 because there's no clear opposite/complement to `Size()`.

Reply:
Would we still want a MaxSize() if using method 2?

Reply:
I would lean toward no, since I don't believe there is a FreeRTOS queue function for it, so we wouldn't be able to put it in the C API as well. But I could be convinced otherwise.